### PR TITLE
Added support multi category

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #8

### CODE REVIEW

File: src/CategorySource.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    private MessageWriterInterface $writer;
+
+    /**
+     * @param string $name Category name.
+     * @param MessageReaderInterface $reader
+     * @param MessageWriterInterface $writer
+     */
+    public function __construct(string $name, MessageReaderInterface $reader, MessageWriterInterface $writer)
+    {
+        if (!preg_match('/^[a-z0-9_-]+$/i', $name)) {
```
Comment (samdark) [inline src/CategorySource.php:27]: Need extra test where category source name doesn't start with `[a-z0-9_-]`. For example, `!test`.

File: src/Extractor.php [hunk 2]
```diff
@@ -71,6 +71,11 @@ public function setOnly(array $only): void
      */
     public function process(string $filesPath, string $defaultCategory, array $languages, OutputInterface $output): void
     {
+        if (!isset($this->categorySources[$defaultCategory])) {
+            $output->writeln('<comment>Default category not found in list of Categories</comment>');
```
Comment (samdark) [inline src/Extractor.php:75]:
```suggestion
            $output->writeln('<comment>Default category was not found in a list of Categories.</comment>');
```

File: src/CategorySource.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            throw new RuntimeException('Category name is invalid. Only letters and numbers are allowed.');
```
Comment (samdark) [inline src/CategorySource.php:28]:
```suggestion
            throw new RuntimeException('Category name is invalid. Only letters, numbers, dash, and underscore are allowed.');
```

File: src/CategorySource.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+namespace Yiisoft\TranslatorExtractor;
+
+use RuntimeException;
+use Yiisoft\Translator\MessageReaderInterface;
+use Yiisoft\Translator\MessageWriterInterface;
+
+/**
+ * Represents message category for read and write messages
+ */
+final class CategorySource
```
Comment (samdark) [inline src/CategorySource.php:14]: Consider using `CategorySource` directly by adding proxy-methods instead of exposing its internals.

**Review by samdark — COMMENTED** [review]
Need docs adjusted additionally to inline comments.

File: src/CategorySource.php (hunk 1 above)
Comment (darkdef) [reply src/CategorySource.php:27]: Mutation in Uppercase (space also incorrect symbol)

File: src/CategorySource.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+declare(strict_types=1);
+
+namespace Yiisoft\TranslatorExtractor;
+
+use RuntimeException;
+use Yiisoft\Translator\MessageReaderInterface;
+use Yiisoft\Translator\MessageWriterInterface;
+
+/**
+ * Represents message category for read and write messages
```
Comment (samdark) [inline src/CategorySource.php:12]:
```suggestion
 * Represents message category for reading and writing messages.
```

File: src/CategorySource.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * @param string $category Category of messages to get.
+     * @param string $locale Locale of messages to get.
+     *
+     * @psalm-return array<string, array<string, string>>
+     *
+     * @return array All messages from category. The format is the following:
+     *
+     * @see \Yiisoft\Translator\MessageReaderInterface::getMessages()
```
Comment (samdark) [inline src/CategorySource.php:51]:
```suggestion
     * @return array All messages from category. The format is
     * the same as in {@see \Yiisoft\Translator\MessageReaderInterface::getMessages()}.
```

File: src/CategorySource.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    public function readMessages(string $category, string $locale): array
+    {
+        return $this->reader->getMessages($category, $locale);
+    }
+
+    /**
+     * @param string $category
+     * @param string $locale
+     * @param array $messages
```
Comment (samdark) [inline src/CategorySource.php:61]:
```suggestion
     * @param string $category Category to write messages to.
     * @param string $locale Locale to write messages for.
     * @param array $messages Messages to write.
```

**Review by samdark — APPROVED** [review]

**Review by xepozz — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**